internal/npm_install/npm_install.bzl

```diff
@@ -813,25 +813,43 @@ def _yarn_install_impl(repository_ctx):
     is_windows_host = is_windows_os(repository_ctx)
     node = repository_ctx.path(get_node_label(repository_ctx))
     yarn = get_yarn_label(repository_ctx)
+    result = repository_ctx.execute(
+        [repository_ctx.path(yarn), "--version"],
+        timeout = repository_ctx.attr.timeout,
+        quiet = repository_ctx.attr.quiet,
+    )
+    if result.return_code:
+        fail("yarn --version failed: %s (%s)" % (result.stdout, result.stderr))
+    if result.stdout.startswith("1."):
+        yarn_version = "classic"
+    else:
+        yarn_version = "berry"
 
     yarn_args = []
 
     # Set frozen lockfile as default install to install the exact version from the yarn.lock
     # file. To perform an yarn install use the vendord yarn binary with:
     # `bazel run @nodejs//:yarn install` or `bazel run @nodejs//:yarn install -- -D <dep-name>`
     if repository_ctx.attr.frozen_lockfile:
-        yarn_args.append("--frozen-lockfile")
+        if yarn_version == "classic":
+            yarn_args.append("--frozen-lockfile")
+        else:
+            fail("--frozen-lockfile should not be used with yarn 2+. Just pass arguments like --immutable.")
 
     if not repository_ctx.attr.use_global_yarn_cache:
         yarn_args.extend(["--cache-folder", str(repository_ctx.path("_yarn_cache"))])
-    else:
+    elif yarn_version == "classic":
         # Multiple yarn rules cannot run simultaneously using a shared cache.
         # See https://github.com/yarnpkg/yarn/issues/683
         # The --mutex option ensures only one yarn runs at a time, see
         # https://yarnpkg.com/en/docs/cli#toc-concurrency-and-mutex
         # The shared cache is not necessarily hermetic, but we need to cache downloaded
         # artifacts somewhere, so we rely on yarn to be correct.
         yarn_args.extend(["--mutex", "network"])
+    else:
+        # Can't tell from documentation if Yarn Berry has any replacement for the --mutex
+        # flag. We'll have to assume it's safe to run concurrently.
```

Comment:
There is no replacement for it, we fixed the cache so it could be atomic.

Reply:
> Can't tell from documentation if Yarn Berry has any replacement for the --mutex flag.

My observation on this:

There are a lot of knobs being exposed through `.yarnrc` file https://yarnpkg.com/configuration/yarnrc#enableGlobalCache so long-term it's probably best if we stop exposing each individual flags through the attributes of `yarn_install` rule.

Instead, we can create a `yarnrc_file` attribute to raise awareness to folks that they should be passing the config file explicitly to the rule. Underneath implementation, we just merge that with `data` and retain current behavior.

Reply:
That sounds like a good feature to me. However I don't see how to pass the config file path to `yarn install` - it looks like it's just automatically inferred from its location. Maybe this could be contributed separately. (maybe it already works just by passing `.yarnrc.yml` to the `data` of yarn_install and then it's auto-detected?)

Reply:
Yup, my understanding is that folks are already passing it through data and yarn auto detects it inside the sandbox.

Reply:
[yarn classic](https://github.com/yarnpkg/yarn/blob/300c47c336be0e8dbc722ef451148a79d9a35486/src/cli/index.js#L75) (v1) has flag called `--use-yarnrc` to specify the yarnrc location.

[yarn berry](https://github.com/yarnpkg/berry/blob/ce1e0c31781e3fc791cfa86a01348c4f0dba9aec/packages/yarnpkg-core/sources/Configuration.ts#L1133) searches for the rc file automatically and does not have this flag, but you could change the filename of the rc file to search for by setting the [environment variable](https://github.com/yarnpkg/berry/blob/ce1e0c31781e3fc791cfa86a01348c4f0dba9aec/packages/yarnpkg-core/sources/Configuration.ts#L869) `YARN_RC_FILENAME`